### DISCUSSION TITLE
MotionCommander: update rate of the command

### DISCUSSION THREAD

**Discussion by alexandre-bonnefond** [bitcraze/discussions#1116] [Q&A] [answered]
Hi there, 

I am building a "flocking controller" for a swarm a crazyflies and I use the Lightghouse system for positioning. The idea is that I compute some repulsive and attractive forces based on the positions and velocities of the CF and feed these as velocity inputs using the `start_linear_motion(vx, vy, vz)` command. Everything seems to work well when the average velocity of the swarm is low but whenever it is high (due to position initialisation or noises) it appears that the system struggles to behave as it should even though the printing function displays the appropriate behavior. 

I think it has to do with the rate at which commands are updated but I am not sure how to verify it. I also saw in the `start_linear_motion(vx, vy, vz)` command that the velocity input is actually put in some sort of _queue_ which would explain the wrong behavior I get as the CF are still computing previous commands delayed from the actual situation. I'll put relevant parts of the code here for a better understanding:

```
import time
import numpy as np
import cflib.crtp
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm
from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.positioning.motion_commander import MotionCommander


pos_dict = {}
vel_dict = {}

def compute_distance(pos1, pos2):
    return np.linalg.norm(pos1 - pos2)

def log_callback(uri, timestamp, data, logconf):
    x = data['stateEstimate.x']
    y = data['stateEstimate.y']
    z = data['stateEstimate.z']
    pos = np.array([x, y, z])
    pos_dict[uri] = pos

    vx = data['stateEstimate.vx']
    vy = data['stateEstimate.vy']
    vz = data['stateEstimate.vz']
    vel = np.array([vx, vy, vz])
    vel_dict[uri] = vel

def start_states_log(scf):
    log_conf = LogConfig(name='States', period_in_ms=100)
    log_conf.add_variable('stateEstimate.x', 'float')
    log_conf.add_variable('stateEstimate.y', 'float')
    log_conf.add_variable('stateEstimate.z', 'float')

    log_conf.add_variable('stateEstimate.vx', 'float')
    log_conf.add_variable('stateEstimate.vy', 'float')
    log_conf.add_variable('stateEstimate.vz', 'float')
    uri = scf.cf.link_uri
    scf.cf.log.add_config(log_conf)
    log_conf.data_received_cb.add_callback(lambda timestamp, data, logconf: log_callback(uri, timestamp, data, logconf))
    log_conf.start()

def start_distance_printing(scf):

    for i in pos_dict.keys():
        if i == scf.cf.link_uri: continue
        dist = compute_distance(pos_dict[scf.cf.link_uri], pos_dict[i])

        print('Distance between URI {} and URI {} is {}'.format(i, scf.cf.link_uri, dist))
    print("\n")



uris = [
    'radio://0/80/2M/E7E7E7E701',
    'radio://0/80/2M/E7E7E7E702',
    'radio://0/80/2M/E7E7E7E703',
]

def compute_repulsion_force(positions, whichCF, r_rep, p_rep, normalize = False):
    [...]
    return repulsion_force

def compute_attraction_force(positions, whichCF, r_att, p_att, normalize = False):
   [...]
    return attraction_force

def compute_self_propulsion(velocities, whichCF, v_0):
   [...]
    return spp_force

def is_in_box_limit(box_limits, positions, whichCF):

    x = positions[whichCF][0]
    y = positions[whichCF][1]
    z = positions[whichCF][2]

    if abs(x) >= box_limits[0]:
        print("x limit reached")
        return False
        
    
    if abs(y) >= box_limits[1]:
        print("y limit reached")
        return False

    if abs(z) >= box_limits[2]:
        print("z limit reached")
        return False

    return True

def run_sequence(scf: SyncCrazyflie):

    start_time = time.time()
    current_time = start_time
    total_time = 20
    box_limits = [1.0, 1.0, 1.0]

    with MotionCommander(scf) as mc:
        time.sleep(3)
        while current_time < total_time + start_time:

            start_distance_printing(scf)

            if is_in_box_limit(box_limits, pos_dict, scf.cf.link_uri):
                att = compute_attraction_force(pos_dict, scf.cf.link_uri, 0.45, 0.4, False)
                rep = compute_repulsion_force(pos_dict, scf.cf.link_uri, 0.3, 0.6, False)
                spp = compute_self_propulsion(vel_dict, scf.cf.link_uri, 0.3)

                force = att + rep + spp
                mc.start_linear_motion(force[0], force[1], force[2])

                time.sleep(0.1)
            
            else:
                mc.stop()
                time.sleep(0.1)

            current_time = time.time()


if __name__ == '__main__':
    cflib.crtp.init_drivers()
    factory = CachedCfFactory(rw_cache='./cache')
    with Swarm(uris, factory=factory) as swarm:
        print('Connected to  Crazyflies')
        swarm.reset_estimators()
        print('Estimators have been reset')
        swarm.parallel_safe(start_states_log)
        print('Logging states info...')
        swarm.parallel_safe(run_sequence)
```

So my question is, does the  `mc.start_linear_motion(force[0], force[1], force[2])` clears the previous command and instantaneoulsy runs the new one and if not how can I manage to do that ? Also, could the _LogConfig_ rate be responsible of that problem ? 

Thank you so much for your help !

**Comment by knmcguire** [comment, marked as answer]
Hi! 

The current update period of the motion commander is on 0.2 seconds, which can be changed upon [initialization of the thread](https://github.com/bitcraze/crazyflie-lib-python/blob/ffd4642b4be1187856c89bb217f883e11330e45f/cflib/positioning/motion_commander.py#L422). That is a bit fixed now so if you'd like to chance it to something else you'd need to [install the python library from source ](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/installation/install/#install-from-source) and change it in the code.

The logconfig rate is at 100 ms so that doesn't seem much. However, one crazyradio to the crazyflies have a packet limitation of about 400 packets per second. Moreover, there are 3 threads of the swarm class, of which each start another thread of the motioncommander, which means that there are 3 queues from three threads that the crazyradio needs to handle, and I'm not sure if the motion commander is suitable for swarms because of this...

Since you are flying with multiple I would normally recommend you to use the high-level commander, but since you are using velocity commands that is a bit tricky as there is no full support for this yet (see this [ticket](https://github.com/bitcraze/crazyflie-firmware/issues/1339)). So I would rather advise to use the `send_setpoint_hover(`) from the [commander class](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/commander/). This requires you to send the setpoint at a regular interval (which you already do), or else a timeout will occur on the Crazyflie side. The [swarmSequence example ](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/swarm/swarmSequence.py
) shows an example of that setpoint sending as well.


Hope this helps! Btw, I think that the topic of[ our latest blogpost](https://www.bitcraze.io/2024/01/the-commander-framework-part-2-offboard-or-onboard/) might be of interest to you :)

**Reply by alexandre-bonnefond** [reply]
Hi !

Thank you for your answer. As you suggested I switched to the commander class but using `send_velocity_world_setpoint()` as it seemed more suited to my case and now everything works perfectly. I manage to have a nice uniform movement witht 3 CF. 

Now that I know that it works with cflib I'll try to switch to Crazyswarm2 and see if I can implement a fully decentralized model. 

I am also interested about the progress of this [ticket](https://github.com/bitcraze/crazyflie-firmware/issues/1339) and I was wondering if I can try to contribute or is it something only done internally at Bitcraze ? 

Thanks again and have a great day !

**Reply by knmcguire** [reply]
Great!

Currently there is no progress on that ticket and there won't be for next quarter I think, but if you'd like to share your view on it and perhaps try something out that can be used for a pull request, ofcourse be my guest :) But perhaps before you start coding, in the ticket itself let's discuss the right approach.
